pkg/template/controller/templateinstance_finalizer.go

```diff
@@ -0,0 +1,259 @@
+				c.enqueue(t)
+			}
+		},
+	})
+
+	var err error
+	c.client, err = dynamic.NewForConfig(c.config)
+	if err != nil {
+		glog.Errorf("failure creating dynamic client: %v", err)
+		return nil
+	}
+
+	return c
+}
+
+// getTemplateInstance returns the TemplateInstance from the shared informer,
+// given its key (dequeued from c.queue).
+func (c *TemplateInstanceFinalizerController) getTemplateInstance(key string) (*templateapi.TemplateInstance, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.lister.TemplateInstances(namespace).Get(name)
+}
+
+// sync is the actual controller worker function.
+func (c *TemplateInstanceFinalizerController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstance(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if templateInstance.DeletionTimestamp == nil {
+		return nil
+	}
+
+	needsFinalizing := false
+	for _, v := range templateInstance.Finalizers {
+		if v == templateapi.TemplateInstanceFinalizer {
+			needsFinalizing = true
+			break
+		}
+	}
+	if !needsFinalizing {
+		return nil
+	}
+
+	glog.V(4).Infof("TemplateInstanceFinalizer controller: syncing %s", key)
+
+	errs := []error{}
+	foreground := metav1.DeletePropagationForeground
+	deleteOpts := &metav1.DeleteOptions{PropagationPolicy: &foreground}
+	for _, o := range templateInstance.Status.Objects {
+		glog.V(5).Infof("attempting to delete object: %#v", o)
+
+		gv, err := schema.ParseGroupVersion(o.Ref.APIVersion)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing group version %s for object %#v: %v", o.Ref.APIVersion, o, err))
+			continue
+		}
+		gk := schema.GroupKind{
+			Group: gv.Group,
+			Kind:  o.Ref.Kind,
+		}
+
+		// if a resource type is removed, the template instance finalizer will
+		// never be able to clean up the template instance since it won't be
+		// able to map+delete all child resources that were previously created.
+		mapping, err := c.dynamicRestMapper.RESTMapping(gk, gv.Version)
+		if err != nil || mapping == nil {
+			errs = append(errs, fmt.Errorf("error mapping object %#v: %v", o, err))
+			continue
+		}
+
+		namespaced := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		namespace := ""
+		if namespaced {
+			namespace = o.Ref.Namespace
+		}
+
+		gvr := schema.GroupVersionResource{
+			Group:    gv.Group,
+			Version:  gv.Version,
+			Resource: mapping.Resource,
+		}
+
+		var resourceClient dynamic.DynamicResourceInterface
+		if namespaced {
+			resourceClient = c.client.NamespacedResource(gvr, namespace)
+		} else {
+			resourceClient = c.client.ClusterResource(gvr)
+		}
+		err = resourceClient.Delete(o.Ref.Name, deleteOpts)
+		if err != nil && !errors.IsNotFound(err) {
+			errs = append(errs, fmt.Errorf("error deleting object %#v with mapping %#v: %v", o, mapping, err))
+			continue
+		}
+	}
+	if len(errs) > 0 {
+		return kerrs.NewAggregate(errs)
```

Comment:
@liggitt do we allow status to be updated on objects being finalized?

Reply:
afaik, all updates (both to spec and to status) are allowed